### PR TITLE
WIP: Reimplement Steel Battalion controller

### PR DESCRIPTION
**Pull request by LanHikariDS**
The end-goal should be self-explanatory, cleaning up the old VSBC code and updating it for the new input system
Still a work in progress, as I can't figure out how to set everything properly (InputManager.h being the current issue). Also my first time working in quite a few different areas (Github, Visual Studio, and (to my knowledge) C++), so there are bound a few mistakes and superfluous commits

### CODE REVIEW

File: src/common/input/InputManager.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ int Gui2XboxPortArray[4] = {
 int dev_num_buttons[to_underlying(XBOX_INPUT_DEVICE::DEVICE_MAX)] = {
 	XBOX_CTRL_NUM_BUTTONS, // MS_CONTROLLER_DUKE
 	XBOX_CTRL_NUM_BUTTONS, // MS_CONTROLLER_S
-	0,
+	XBOX_CTRL_NUM_BUTTONS, // STEEL_BATTALION_CONTROLLER This probably isn't correct?? Let's see what happens
```
Comment (RadWolfie) [inline src/common/input/InputManager.cpp:63]: Please move this change down to line 67 (index 6 which is STEEL_BATTALION_CONTROLLER originally located).

File: src/common/input/InputManager.cpp [hunk 2]
```diff
@@ -347,16 +347,16 @@ bool InputDeviceManager::UpdateXboxPortInput(int usb_port, void* Buffer, int Dir
 				switch (xid_type)
 				{
 				case to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_DUKE):
-                case to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S): {
+                case to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S):
+				case to_underlying(XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER): {
```
Comment (RadWolfie) [inline src/common/input/InputManager.cpp:351]: I am unsure about this change since maybe there is a way to able dynamic support different virtual controller within `UpdateInputXpad`? If not, it will need a brand new function such as `UpdateInputSBC`.

File: src/common/input/InputManager.h [hunk 3]
```diff
@@ -62,11 +63,25 @@ struct XpadInput {
 	int16_t sThumbRY;
 };
 
+struct SBCInput { //I'm not quite sure how this works, but it should come out to 0x18
```
Comment (RadWolfie) [inline src/common/input/InputManager.h:66]: Please move SBCInput down to before SBCOutput structure is at or the other way around for SBCOutput structure move up. We're not sorting base on Input/Output, we're sorting base on controller type listing.

File: src/devices/usb/XidGamepad.cpp [hunk 4]
```diff
@@ -242,7 +274,7 @@ XboxDeviceState* XidGamepad::ClassInitFn()
 		m_pPeripheralFuncStruct->handle_destroy = std::bind(&XidGamepad::UsbXid_HandleDestroy, this);
 		m_pPeripheralFuncStruct->handle_attach  = std::bind(&XidGamepad::UsbXid_Attach, this, _1);
 		m_pPeripheralFuncStruct->product_desc   = dev->ProductDesc.c_str();
-		m_pPeripheralFuncStruct->usb_desc       = &desc_xbox_gamepad;
+		m_pPeripheralFuncStruct->usb_desc       = &desc_SBC_gamepad; //Put split for SBC controller type
```
Comment (RadWolfie) [inline src/devices/usb/XidGamepad.cpp:277]: Since this is only for testing, please remember to revert this change.

**Comment by RadWolfie** [comment]
**Friendly reminder for anyone trying out this pull request build(s):**

It is intended for ongoing experiment to add Steel Battalion controller support for specific titles that support it. It is very likely other titles will not work as intended. Please refrain report issues with other titles until pull request has move out of draft status.
Thanks!

File: src/common/input/InputManager.cpp (hunk 1 above)
Comment (ergo720) [reply src/common/input/InputManager.cpp:63]: The correct number to use is documented in EmuShared.h
```
char m_DeviceControlNames[4][XBOX_CTRL_NUM_BUTTONS][30]; // macro should be num of buttons of dev with highest num buttons
```

File: src/common/input/InputManager.cpp (hunk 2 above)
Comment (ergo720) [reply src/common/input/InputManager.cpp:351]: This code change is wrong: `UpdateInputXpad` is specific for xbox duke and S controllers. The xboxdevwiki documents that the steel battalion controller uses totally different buttons, so the function can't work with those and thus this requires a new function specifically for this controller type.

File: src/common/input/InputManager.h (hunk 3 above)
Comment (ergo720) [reply src/common/input/InputManager.h:66]: This struct should represent the number of input buttons that the steel battalion controller uses. These are documented in the xboxdevwiki in the section "Controller to Xbox" and you can see that they don't correspond to those used by standard duke and S controller, which is what this struct, as written here, is using. Thus this needs changing as well.

File: src/common/input/InputManager.h [hunk 5]
```diff
@@ -62,11 +63,25 @@ struct XpadInput {
...
+	int16_t sThumbRY;
+};
+
 struct XpadOutput {
 	uint16_t left_actuator_strength;
 	uint16_t right_actuator_strength;
 };
 
+struct SBCOutput { //Likewise, with 0x14
+	uint16_t left_actuator_strength;
```
Comment (ergo720) [inline src/common/input/InputManager.h:81]: These should be the output controls of the steel battalion controller, which are documented in the xboxdevwiki in the section "Xbox to Controller".

File: src/common/input/InputManager.h (hunk 3 above)
Comment (LanHikariDS) [reply src/common/input/InputManager.h:66]: > This struct should represent the number of input buttons that the steel battalion controller uses. These are documented in the xboxdevwiki in the section "Controller to Xbox" and you can see that they don't correspond to those used by standard duke and S controller, which is what this struct, as written here, is using. Thus this needs changing as well.
I'm aware this is incorrect, I asked about it on Discord and got no response, so I wasn't sure how to define them, hence the comment

File: src/core/hle/XAPI/Xapi.cpp [hunk 6]
```diff
@@ -80,7 +80,8 @@ bool operator==(xbox::PXPP_DEVICE_TYPE XppType, XBOX_INPUT_DEVICE XidType)
 	switch (XidType)
 	{
 	case XBOX_INPUT_DEVICE::MS_CONTROLLER_DUKE:
-    case XBOX_INPUT_DEVICE::MS_CONTROLLER_S: {
+    case XBOX_INPUT_DEVICE::MS_CONTROLLER_S:
+    case XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER: {
```
Comment (ergo720) [inline src/core/hle/XAPI/Xapi.cpp:84]: The device type used by the steel battalion controller is different than the one used by xbox gamepads. We are deriving it in `SetupXboxDeviceTypes` at line 227:
```
                case XINPUT_DEVTYPE_STEELBATALION:
                    printf("XDEVICE_TYPE_STEELBATALION)\n");
                    break;
```

So, this required a new global xpp type for this controller type.

File: src/core/hle/XAPI/Xapi.cpp [hunk 7]
```diff
@@ -140,7 +140,17 @@ bool ConstructHleInputDevice(int Type, int Port)
     }
 	break;
 	case to_underlying(XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER): {
-		// TODO
+		g_XboxControllerHostBridge[Port].XboxPort = Port;
+        g_XboxControllerHostBridge[Port].XboxType = XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER;
+        g_XboxControllerHostBridge[Port].InState = new XpadInput();
```
Comment (ergo720) [inline src/core/hle/XAPI/Xapi.cpp:145]: `XpadInput` is specific is for xbox controllers, so this should use `SBCInput` instead.

File: src/devices/usb/XidGamepad.cpp (hunk 4 above)
Comment (LanHikariDS) [reply src/devices/usb/XidGamepad.cpp:277]: I don't know how to reference the current controller's type in this part of the code

File: src/devices/usb/XidGamepad.cpp [hunk 8]
```diff
@@ -203,6 +203,38 @@ static const XIDDesc desc_xid_xbox_gamepad = {
...
+		0x0a7b,            // idVendor
+		0xd000,            // idProduct
+		0x0100,            // bcdDevice
+		STR_MANUFACTURER,  // iManufacturer
+		STR_PRODUCT,       // iProduct
+		STR_SERIALNUMBER   // iSerialNumber
+	},
+	&desc_device_SBC_gamepad
+};
+
```
Comment (ergo720) [inline src/devices/usb/XidGamepad.cpp:237]: Ni: the USB descriptors are only necessary for LLE USB, you don't need those for HLE XInput so these additions are unnecessary but probably fine to have.

File: src/devices/usb/XidGamepad.cpp (hunk 4 above)
Comment (ergo720) [reply src/devices/usb/XidGamepad.cpp:277]: `XidGamepad` is a class which implements an LLE xbox duke controller, so this change is wrong and should be reverted.

File: src/common/input/InputDevice.h [hunk 9]
```diff
@@ -58,11 +58,11 @@ typedef enum class _XBOX_INPUT_DEVICE : int {
 	DEVICE_INVALID = -1,
 	MS_CONTROLLER_DUKE,
 	MS_CONTROLLER_S,
+	STEEL_BATTALION_CONTROLLER,
```
[the remark this answers is not in the document]
Comment (LanHikariDS) [reply src/common/input/InputDevice.h:61]: I didn't see any need to revert it since I'd already changed it, but sure

File: src/gui/DlgInputConfig.cpp [hunk 10]
```diff
@@ -188,6 +188,12 @@ INT_PTR CALLBACK DlgInputConfigProc(HWND hWndDlg, UINT uMsg, WPARAM wParam, LPAR
 				}
 				break;
 
+				case to_underlying(XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER): {
+					DialogBoxParam(GetModuleHandle(nullptr), MAKEINTRESOURCE(IDD_VIRTUAL_SBC_FEEDBACK), hWndDlg, DlgXidControllerConfigProc,
+						(DeviceType << 8) | port);
```
Comment (ergo720) [inline src/gui/DlgInputConfig.cpp:193]: `DlgXidControllerConfigProc` will open the input dialog for an xbox controller, while I imagine you want to open the old input dialog for the steel battalion controller. Thus this requires a new function to handle the old input dialog.

File: src/devices/usb/XidGamepad.cpp (hunk 8 above)
Comment (RadWolfie) [reply src/devices/usb/XidGamepad.cpp:237]: Yet it will be great to have SB controller prepped for LLE USB support. 😉 Beside, I do recall Luke need go down full LLE USB in order to emulate Chihiro system. Or isn't using our LLE USB manager for variant of controllers on retail Xbox's end.

Comment (ergo720) [reply src/devices/usb/XidGamepad.cpp:237]: Except that I just remembered that XidGamepad is specific for duke controllers only, so the descriptors for the steel battalion should really go in its own file which implements the corresponding LLE device. Thus, putting them here feels incorrect to me.

Comment (LanHikariDS) [reply src/devices/usb/XidGamepad.cpp:237]: > 
> 
> Nit: the USB descriptors are only necessary for LLE USB, you don't need those for HLE XInput so these additions are unnecessary but probably fine to have.

Noted, not quite sure what to do with it, then...

File: src/gui/DlgInputConfig.cpp (hunk 10 above)
Comment (LanHikariDS) [reply src/gui/DlgInputConfig.cpp:193]: But I've already got the old SBC input dialogue opening. Of course, the old dialogue isn't the final solution, as I'll need to redo it for bindings and such, but is that the only issue with this section?

**Comment by RadWolfie** [comment]
@LanHikariDS Would you like me to send you a pull request with new files for you to work on? There may will be more than one pull request to assist you on the right path. Though, I'm not able to assist right now. Some time after 24 hours from now I will.

File: src/core/hle/XAPI/Xapi.cpp (hunk 6 above)
Comment (ergo720) [reply src/core/hle/XAPI/Xapi.cpp:84]: To be honest, I'm not totally sure if the xpp type of the steel battalion controller is different than that of xbox gamepads. If it is the same, then `g_DeviceType_Gamepad` can be reused here. However, even in that case, the current code in `SetupXboxDeviceTypes` simply logs the type but does nothing to it, which means that `g_DeviceType_Gamepad` will stay as nullptr. So, I still suggest the below change if the xpp type is the same:
```
case XINPUT_DEVTYPE_STEELBATALION:
g_DeviceType_Gamepad = deviceTable[i]->XppType;
printf("XDEVICE_TYPE_STEELBATALION)\n");
break;
```

**Comment by LanHikariDS** [comment]
> @LanHikariDS Would you like me to send you a pull request with new files for you to work on? There may will be more than one pull request to assist you on the right path. Though, I'm not able to assist right now. Some time after 24 hours from now I will.

That would probably be helpful, as I'm currentlystruggling with SBCInput and SBCOutput, as XpadInput doessn't match XboxDevWiki, and SBCOutput has some tags on XboxDevWiki that cause issues when compiling

**Comment by ergo720** [comment]
>XpadInput doessn't match XboxDevWiki

It does, look carefully at the offsets and masks reported in the wiki. It says that the first field starts at offset 2, and that's because the first two fields are `bReportId` and `bLength`, which are only relevant for LLE USB and thus are ignored in `XpadInput`. All the following 8 buttons start at the same offset 2 and occupy 2 bytes (because the first button following them starts at offset 4). The masks for these buttons tell you that they are represented by a single bit, and that's because they are digital buttons, that is, they can only be on or off. The following 8 buttons are one byte each and are analog instead, and their intensity values can range between 0 - 255 (0 - 0xFF). Finally, the analog sticks use two bytes each, and their state is represented by a signed 16 bit value (because the wiki says they can have negative values).

**Comment by LanHikariDS** [comment]
> It does, look carefully at the offsets and masks reported in the wiki. It says that the first field starts at offset 2, and that's because the first two fields are `bReportId` and `bLength`, which are only relevant for LLE USB and thus are ignored in `XpadInput`. All the following 8 buttons start at the same offset 2 and occupy 2 bytes (because the first button following them starts at offset 4). The masks for these buttons tell you that they are represented by a single bit, and that's because they are digital buttons, that is, they can only be on or off. The following 8 buttons are one byte each and are analog instead, and their intensity values can range between 0 - 255 (0 - 0xFF). Finally, the analog sticks use two bytes each, and their state is represented by a signed 16 bit value (because the wiki says they can have negative values).

So the actual names (bAnalogButtons[8]) aren't actually important, and don't need to be anything specific?

File: src/core/hle/XAPI/Xapi.cpp (hunk 6 above)
Comment (ObiKKa) [reply src/core/hle/XAPI/Xapi.cpp:84]: Remember the spelling (double T's): "STEELBATTALION"

**Comment by ergo720** [comment]
Using an array is done just for brevity: this way, I don't need to add 8 separate struct members. Also, the SB input struct has already being implemented for you in `Xapi.h` at line 148:
```
typedef struct _X_SBC_GAMEPAD {
    xbox::word_xt    wButtons[3];
    xbox::short_xt   sAimingX;
    xbox::short_xt   sAimingY;
    xbox::short_xt   sRotationLever;//maybe only high byte was used.
    xbox::short_xt   sSightChangeX;
    xbox::short_xt   sSightChangeY;
    xbox::word_xt    wLeftPedal;//maybe only high byte was used.
    xbox::word_xt    wMiddlePedal;//maybe only high byte was used.
    xbox::word_xt    wRightPedal;//maybe only high byte was used.
    xbox::uchar_xt   ucTunerDial;//low nibble, The 9 o'clock postion is 0, and the 6 o'clock position is 12. The blank area between the 6 and 9 o'clock positions is 13, 14, and 15 clockwise.
    xbox::uchar_xt   ucGearLever;//GearLever 1~5 for gear 1~5, 7~13 for gear R,N,1~5, 15 for gear R. we use the continues range from 7~13
}
X_SBC_GAMEPAD, *PX_SBC_GAMEPAD;
```
The function `EmuSBCGetState` in `Xapi.cpp` at line 559 also implements filling this input state from the host controller input state, but please don't just copy it because, like this comment https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/1196#issuecomment-390973441 says, it's using hard-coded bindings, while we want to let users set arbitrary bindings. You can check the function `UpdateInputXpad` to see how I did that for xbox controllers.

Regarding the output state of `SBCOutput`: this represents the LED state of certain buttons of the SB controller. However, to my knowledge, XInput, DInput and SDL don't support LED states on buttons (assuming the host controller even has LEDs in the first place), so you can't really implement the SB output state and so you can just ignore it and only implement the input state.

One thing I notice in the `XInputGetState` patch: it will break with the SB device because the `pState` buffer used by it will not be `Gamepad` but a SB specific buffer, and thus the `memcpy` operation below will be incorrect (`xbox::XINPUT_GAMEPAD` is 18 bytes while `xbox::X_SBC_GAMEPAD` is 24 bytes instead). To handle that, I suggest the below code change:
```
memcpy((void *)&pState->Gamepad, g_XboxControllerHostBridge[Port].InState, g_XboxControllerHostBridge[Port].XboxDeviceInfo.ucInputStateSize);
```

File: src/core/hle/XAPI/Xapi.cpp (hunk 6 above)
Comment (ergo720) [reply src/core/hle/XAPI/Xapi.cpp:84]: Ok, I checked with the game and I can confirm that the xpp type for the SB controller is different, so you need a new global variable here.
```
g_DeviceType_SBC = deviceTable[i]->XppType;
```

@ObiKKa That's an existing typo in the code, I merely copied it as is, but yes it needs to be fixed as well.

**Comment by RadWolfie** [comment]
> Regarding the output state of `SBCOutput`: this represents the LED state of certain buttons of the SB controller. However, to my knowledge, XInput, DInput and SDL don't support LED states on buttons (assuming the host controller even has LEDs in the first place), so you can't really implement the SB output state and so you can just ignore it and only implement the input state.

Seems only accessible through USB level base on google search. SDL apparently only has one LED support depending on how it's implement.

**Comment by LanHikariDS** [comment]
> Regarding the output state of `SBCOutput`: this represents the LED state of certain buttons of the SB controller. However, to my knowledge, XInput, DInput and SDL don't support LED states on buttons (assuming the host controller even has LEDs in the first place), so you can't really implement the SB output state and so you can just ignore it and only implement the input state.
Just trying to have it match up so the Feedback size comes out the same, rather than actually work, as according to Jackchen, it needs to be 0x14. As for the Input struct, currently just trying to get the game to _recognize_ the controller before worry about the buttons working and such, but my end goal is indeed for them to be remappable, not just using hardcoded binds

File: src/core/hle/XAPI/Xapi.cpp [hunk 11]
```diff
@@ -80,14 +81,20 @@ bool operator==(xbox::PXPP_DEVICE_TYPE XppType, XBOX_INPUT_DEVICE XidType)
...
-    case XBOX_INPUT_DEVICE::MS_CONTROLLER_S:
-    case XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER: {
+    case XBOX_INPUT_DEVICE::MS_CONTROLLER_S: {
 		if (XppType == g_DeviceType_Gamepad) {
 			return true;
 		}
 	}
 	break;
 
+	case XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER: {
```
Comment (RadWolfie) [inline src/core/hle/XAPI/Xapi.cpp:91]: Great finding! Although, it is currently not connected to `SetupXboxDeviceTypes` function. You should be able to find keyword `XINPUT_DEVTYPE_STEELBATTALION`. Plus possibility replace `break;` to `continue;` for allow iterate whole table.

File: src/core/hle/XAPI/Xapi.cpp [hunk 12]
```diff
@@ -61,6 +61,7 @@ PFARPROC1 fnCxbxVSBCOpen;
 //typedef DWORD(*fnCxbxVSBCSetState)(UCHAR *);
 //typedef DWORD(*fnCxbxVSBCGetState)(UCHAR *);
 xbox::PXPP_DEVICE_TYPE g_DeviceType_Gamepad = nullptr;
+xbox::PXPP_DEVICE_TYPE g_DeviceType_SBC = nullptr; //deviceTable[i]->XppType?
```
Comment (RadWolfie) [inline src/core/hle/XAPI/Xapi.cpp:64]: Oh for your questionable comment. Yes, that is correct as I referred to previous comment.

File: src/core/hle/XAPI/Xapi.cpp (hunk 11 above)
Comment (ergo720) [reply src/core/hle/XAPI/Xapi.cpp:91]: No need for `continue`, the whole table will still be iterated even with the `break` inside the `switch`.

Comment (RadWolfie) [reply src/core/hle/XAPI/Xapi.cpp:91]: Ah, I didn't see the switch statement. 😅

File: src/common/input/InputManager.h [hunk 13]
```diff
@@ -67,6 +68,74 @@ struct XpadOutput {
 	uint16_t right_actuator_strength;
 };
 
+struct SBCInput1 { //I'm not quite sure how this works, but it should come out to 0x18
```
Comment (ergo720) [inline src/common/input/InputManager.h:71]: This struct is wrong and should just be deleted.

File: src/common/input/InputManager.h [hunk 14]
```diff
@@ -67,6 +68,74 @@ struct XpadOutput {
...
+	uint16_t wButtons;
+	uint8_t bAnalogButtons[8];
+	int16_t sThumbLX;
+	int16_t sThumbLY;
+	int16_t sThumbRX;
+	int16_t sThumbRY;
+};
+
+typedef struct SBCInput { //Was told not to copy this, did it anyways for the sake of having the proper input size and potentially modifying it's bindings from there.
```
Comment (ergo720) [inline src/common/input/InputManager.h:80]: No need for `typedef` here, you can just write `struct SBCInput` + you can remove the `X_SBC_GAMEPAD, * PX_SBC_GAMEPAD;` below. Also, this is the input struct as used by the input system. As such it should not use the xbox types, but just the host types instead. word_xt -> uint16_t, short_xt -> int16_t and uchar_xt -> uint8_t.

File: src/core/hle/XAPI/Xapi.cpp [hunk 15]
```diff
@@ -140,7 +147,17 @@ bool ConstructHleInputDevice(int Type, int Port)
...
+		g_XboxControllerHostBridge[Port].XboxPort = Port;
+        g_XboxControllerHostBridge[Port].XboxType = XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER;
+        g_XboxControllerHostBridge[Port].InState = new SBCInput();
+        g_XboxControllerHostBridge[Port].bPendingRemoval = false;
+        g_XboxControllerHostBridge[Port].bSignaled = false;
+        g_XboxControllerHostBridge[Port].bIoInProgress = false;
+        g_XboxControllerHostBridge[Port].XboxDeviceInfo.ucType = XINPUT_DEVTYPE_STEELBATTALION;
+        g_XboxControllerHostBridge[Port].XboxDeviceInfo.ucSubType = XINPUT_DEVSUBTYPE_GC_GAMEPAD_ALT;
+        g_XboxControllerHostBridge[Port].XboxDeviceInfo.ucInputStateSize = sizeof(SBCInput);
+        g_XboxControllerHostBridge[Port].XboxDeviceInfo.ucFeedbackSize = sizeof(XpadOutput);
```
Comment (ergo720) [inline src/core/hle/XAPI/Xapi.cpp:159]: This should be 0x14 instead of taking the size of the output xbox controller struct.

File: src/common/input/InputManager.h [hunk 16]
```diff
@@ -67,6 +68,74 @@ struct XpadOutput {
...
+	xbox::short_xt   sSightChangeY;
+	xbox::word_xt    wLeftPedal;//maybe only high byte was used.
+	xbox::word_xt    wMiddlePedal;//maybe only high byte was used.
+	xbox::word_xt    wRightPedal;//maybe only high byte was used.
+	xbox::uchar_xt   ucTunerDial;//low nibble, The 9 o'clock postion is 0, and the 6 o'clock position is 12. The blank area between the 6 and 9 o'clock positions is 13, 14, and 15 clockwise.
+	xbox::uchar_xt   ucGearLever;//GearLever 1~5 for gear 1~5, 7~13 for gear R,N,1~5, 15 for gear R. we use the continues range from 7~13
+}
+X_SBC_GAMEPAD, * PX_SBC_GAMEPAD;
+
+struct SBCOutput { //Likewise, with 0x14
```
Comment (ergo720) [inline src/common/input/InputManager.h:95]: Since we are not going to implement the SB output state, this struct is unnecessary and can be deleted. Unless you want to have it for documentation purposes? But there's the xbox wiki already for that, so I'm not sure.

File: src/common/input/InputManager.h (hunk 13 above)
Comment (LanHikariDS) [reply src/common/input/InputManager.h:71]: This was my old attempt at the input struct from scratch, kept because I assumed we wouldn't _actually_ use the copied one in the long run

File: src/core/hle/XAPI/Xapi.cpp (hunk 15 above)
Comment (LanHikariDS) [reply src/core/hle/XAPI/Xapi.cpp:159]: Pretty sure I tried this and it just wouldn't compile, hence why I went through with SBCOutput, but I'll try it again, and resolve your Line 95 comment if it works

File: src/common/input/InputManager.h (hunk 16 above)
Comment (LanHikariDS) [reply src/common/input/InputManager.h:95]: If I can get it to compile and work without it, I don't really see it as valuable for documentation, as it's likely incorrect

**Comment by LanHikariDS** [comment]
Why did this close...?

**Comment by RadWolfie** [comment]
I reopened the pull request for you.

**Comment by RadWolfie** [comment]
Github CI fail to compile, see error message below.
> D:\a\Cxbx-Reloaded\Cxbx-Reloaded\src\core\hle\XAPI\Xapi.cpp(92,26): error C2196: case value '_XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER' already used [D:\a\Cxbx-Reloaded\Cxbx-Reloaded\build\projects\cxbxr-emu\cxbxr-emu.vcxproj]

**Comment by LanHikariDS** [comment]
Yeah, and I see why. the SBC definition got moved, by the old one wasn't cleared

File: src/core/hle/XAPI/Xapi.cpp (hunk 12 above)
Comment (LanHikariDS) [reply src/core/hle/XAPI/Xapi.cpp:64]: Yes, but I'm not sure what you meant by 'deviceTable[i]->XppType'

File: src/core/hle/XAPI/Xapi.cpp (hunk 15 above)
Comment (LanHikariDS) [reply src/core/hle/XAPI/Xapi.cpp]: It does indeed compile, but the game still doesn't recognize the controller...

File: src/core/hle/XAPI/Xapi.cpp (hunk 11 above)
Comment (LanHikariDS) [reply src/core/hle/XAPI/Xapi.cpp:91]: Holy shit, this did it. This finally fucking did it.

**Comment by LanHikariDS** [comment]
For some odd reason, both Steel Battalion games crash when IDC_SET_LEFT_NEGX is added to the input mapping dialog. I tried correcting this by giving it a SB-specific case instead, but that doesn't work either.

File: src/common/input/InputManager.cpp (hunk 2 above)
Comment (LanHikariDS) [reply src/common/input/InputManager.cpp:351]: Was this corrected by your PR, or is this still an issue?

File: src/gui/DlgInputConfig.cpp (hunk 10 above)
Comment (LanHikariDS) [reply src/gui/DlgInputConfig.cpp:194]: I've been having no issues with the UI, but maybe this was an issue before I'd run into it that's now resolved, or do I still need a new function?

**Comment by LanHikariDS** [comment]
Something got fuckered, current build doesn't start

File: src/common/input/InputManager.cpp (hunk 1 above)
Comment (LanHikariDS) [reply src/common/input/InputManager.cpp]: Changing this to SBC_NUM_BUTTONS causes the emulator to crash immediately upon starting up, as seen in https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/2068/commits/0af6f46ade4e6bb98da3590b3f9f5fbd457b2404

Comment (ergo720) [reply src/common/input/InputManager.cpp]: Then you'll have to debug and see why it crashes.

File: src/gui/DlgInputConfig.cpp (hunk 10 above)
Comment (ergo720) [reply src/gui/DlgInputConfig.cpp:194]: This now calls `DlgSBControllerConfigProc`, which is the correct thing to do. Before, you were trying to call `DlgXidControllerConfigProc` for the SBC, which was wrong.

File: src/common/input/InputManager.cpp (hunk 2 above)
Comment (ergo720) [reply src/common/input/InputManager.cpp:351]: No, it wasn't and this is still wrong. You need a new function `UpdateInputSBC` specific for the SBC.

File: src/gui/DlgSBControllerConfig.cpp [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+		int port_num = lParam & 0xFF;
+		int dev_type = (lParam & 0xFF00) >> 8;
+
+		// Ensure that port_num is a valid xbox port
+		assert(port_num >= PORT_1 && port_num <= PORT_4);
+
+		// Ensure that the controller type is valid
+		assert(dev_type == to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_DUKE) ||
+			dev_type == to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S) ||
+			dev_type == to_underlying(XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER));
```
Comment (ergo720) [inline src/gui/DlgSBControllerConfig.cpp:51]: `DlgSBControllerConfigProc` should only assert that the type is "SBC" and not "xbox controller".

File: src/gui/DlgDukeControllerConfig.cpp [hunk 18]
```diff
@@ -50,7 +50,8 @@ INT_PTR CALLBACK DlgXidControllerConfigProc(HWND hWndDlg, UINT uMsg, WPARAM wPar
 
 		// Ensure that the controller type is valid
 		assert(dev_type == to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_DUKE) ||
-			dev_type == to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S));
+			dev_type == to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S) ||
+			dev_type == to_underlying(XBOX_INPUT_DEVICE::STEEL_BATTALION_CONTROLLER));
```
Comment (ergo720) [inline src/gui/DlgDukeControllerConfig.cpp:54]: `DlgXidControllerConfigProc` should only assert that the type is "xbox controller" and not "SBC", so please revert this change.

File: src/gui/DlgSBControllerConfig.cpp (hunk 17 above)
Comment (LanHikariDS) [reply src/gui/DlgSBControllerConfig.cpp:51]: So lines 50 and 51 should be removed from this file, and like 54 should be removed from DlgDukeControllerConfig.cpp, yes?

Comment (RadWolfie) [reply src/gui/DlgSBControllerConfig.cpp:51]: @LanHikariDS, "xbox controller" are default type of controller which are Duke and S controllers. Only thing should be in here is "Steel Battalion Controller". Plus Xbox's Duke controller file at line 54 need to be remove, aka revert, yes.

Duke controller is at line 49. Plus S controller is at line 50.

Asserts are only there to verify the code should be working as intended design in debug build.

Comment (LanHikariDS) [reply src/gui/DlgSBControllerConfig.cpp:51]: These have been corrected on my local end, will commit and push later tonight after rebasing

**Comment by LanHikariDS** [comment]
Aaaaand, I forgot to commit _before_ pushing the rebase and new changes, so it closed again

**Comment by ergo720** [comment]
Superseded by https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/2134
